src/libcore/task.rs

```diff
@@ -467,7 +469,6 @@ impl fmt::Debug for TaskObj {
 }
 
 unsafe impl Send for TaskObj {}
-unsafe impl Sync for TaskObj {}
```

Comment:
Why remove this? We don't particularly *need* it, but it's totally sound AFAICT, and it will give the correct behavior were we to add more non-mutating methods in the future (in addition to the current `Debug::fmt`).

Reply:
I can put it in again, but is it safe? `trait UnsafeTask: Send + 'static ` <-- no `Sync`

Reply:
According to the comment above `TaskObj` it is supposed to be approximately equivalent to `Box<Future<Output = ()> + Send>` which has also no `Sync` bound.

Reply:
`UnsafeTask` only has `&mut` methods-- those are only used for mutating the inner task. Since no `&self` methods are called, it's trivially `Sync`.

Reply:
`TaskObj` is just a workaround for `Box<Future<Output = ()> + Send + 'static>` which throws an error, but if it worked, it wouldn't be be `Sync`. I think the workaround should match what we really want but can't get as much as possible. IMO it feels wrong to implement `Sync`. I agree with your reasoning why it can be considered safe, though.

Here's the error for `Box<Future<Output = ()> + Send + 'static>`:
```rust
let boxed: Box<Future<Output = ()> + Send + 'static> = ...
// error[E0038]: the trait `std::future::Future` cannot be made into an object
// note: method `poll` has a non-standard `self` type
```

Also a [playground example](https://play.rust-lang.org/?gist=240fe6646b6fe20cea257f0fbcef2afd&version=nightly&mode=debug) that shows `Box<Future<Output = ()> + Send + 'static>` is not `Sync`

I believe the more conservative approach would be to leave it out.

Reply:
That is certainly the more conservative approach :).

I'm fine with leaving it out if you feel strongly about it, but I don't think it really matters significantly either way. Currently the only `&self` method is `Debug::fmt`, which is thread-safe, but if we someday decide to add a thread-locked `&self` method you're right that we wouldn't be able to do so.

Reply:
No strong feelings, because, as you say, it's safe. My recommendation is, however, to leave it out.

**For adding `impl Sync`:** Because we can. It's safe.
**Against adding `impl Sync`:** More conservative, we don't need it, and it approximates `Box<Future...>` which is `!Sync` more closely
=> My conclusion: Don't add it